### PR TITLE
Initial implementation of agent build container Dockerfiles

### PR DESCRIPTION
**Pull request by dhrupadb**

### CODE REVIEW

File: build/x86_64_centos6/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Image name stackdriver-logging-rhel6-build-image
+# To push to GCR
```
[the remark this answers is not in the document]
Comment (dhrupadb) [reply build/x86_64_centos6/Dockerfile]: Done.

**Comment by dhrupadb** [comment]
Friendly Ping.

File: build/x86_64_centos6/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ENV PATH ${PATH}:/usr/local/bin
+
+RUN rpm -ivh http://repoforge.mirror.constant.com/redhat/el6/en/x86_64/rpmforge/RPMS/rpmforge-release-0.5.3-1.el6.rf.x86_64.rpm \
+  && yum --enablerepo=rpmforge-extras install -y git \
+  && gpg --keyserver hkp://keys.gnupg.net --recv-keys 409B6B1796C275462A1703113804BB82D39DC0E3 \
+  && git config --global user.email "stackdriver-github-reader@google.com" \
+  && git config --global user.name "stackdriverreader"
+
+RUN curl -sSL https://get.rvm.io | /bin/bash -s stable
+RUN /bin/bash -l -c "rvm requirements && rvm install 2.2 && gem install bundler --no-ri --no-rdoc && gem update"
```
Comment (qingling128) [inline build/x86_64_centos6/Dockerfile]: We could use Ruby 2.4 now instead of 2.2, right? Same for the other two files.

Comment (dhrupadb) [reply build/x86_64_centos6/Dockerfile]: Done.

File: build/x86_64_debian/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,18 @@
+FROM debian:7
+
+#RUN rm /bin/sh && ln -s /bin/bash /bin/sh
+RUN /bin/bash -l -c "apt-get update -y && apt-get install -y \
```
Comment (bmoyles0117) [inline build/x86_64_debian/Dockerfile]: Why are we running this inside of /bin/bash? Same comment everywhere else.

File: build/x86_64_debian/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,18 @@
+FROM debian:7
```
Comment (bmoyles0117) [inline build/x86_64_debian/Dockerfile]: This file name doesn't declare that it's for debian7, should we be consistent with what we're doing for centos?

File: build/x86_64_debian/Dockerfile (hunk 3 above)
Comment (dhrupadb) [reply build/x86_64_debian/Dockerfile]: apt needs a login shell to get rvm to work correctly. This isn't an issue for yum on the other hand.

File: build/x86_64_debian/Dockerfile (hunk 4 above)
Comment (dhrupadb) [reply build/x86_64_debian/Dockerfile]: I tried to keep the containers as closely linked to the repos where the packages are published. There's just one generic debian repo for all logging agent version, which is why I didn't explicitly call out "debian7" or "wheezy" in the container name or directory.

File: build/x86_64_centos6/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  bzip2 \
+  rpm-build \
+  rpm-sign \
+  autoconf \
+  expect \
+  centos-release-SCL
+
+ENV PATH ${PATH}:/usr/local/bin
+
+RUN rpm -ivh http://repoforge.mirror.constant.com/redhat/el6/en/x86_64/rpmforge/RPMS/rpmforge-release-0.5.3-1.el6.rf.x86_64.rpm \
```
Comment (igorpeshansky) [inline build/x86_64_centos6/Dockerfile]: These files seem to have a lot more layers than the Debian one. Any reason we can't fold all of these RUN commands into one?

File: build/x86_64_debian/Dockerfile (hunk 3 above)
Comment (dhrupadb) [reply build/x86_64_debian/Dockerfile]: Actually nvm my bad. We don't need a login shell to install the dependencies. Fixed.

File: build/x86_64_centos6/Dockerfile (hunk 5 above)
Comment (dhrupadb) [reply build/x86_64_centos6/Dockerfile]: Combined them. PTAL

File: build/x86_64_centos6/Dockerfile [hunk 6]
```diff
@@ -15,15 +17,12 @@ RUN yum update -y && yum install -y \
...
-
-RUN rpm -ivh http://repoforge.mirror.constant.com/redhat/el6/en/x86_64/rpmforge/RPMS/rpmforge-release-0.5.3-1.el6.rf.x86_64.rpm \
+  centos-release-SCL \
+  && rpm -ivh http://ftp.tu-chemnitz.de/pub/linux/dag/redhat/el6/en/x86_64/rpmforge/RPMS/rpmforge-release-0.5.3-1.el6.rf.x86_64.rpm \
   && yum --enablerepo=rpmforge-extras install -y git \
   && gpg --keyserver hkp://keys.gnupg.net --recv-keys 409B6B1796C275462A1703113804BB82D39DC0E3 \
   && git config --global user.email "stackdriver-github-reader@google.com" \
-  && git config --global user.name "stackdriverreader"
+  && git config --global user.name "stackdriverreader" \
+  && curl -sSL https://get.rvm.io | /bin/bash -s stable
```
Comment (igorpeshansky) [inline build/x86_64_centos6/Dockerfile]: Careful! This pipe will probably not do what you expect. Let's put the pipe in a subshell...

File: build/x86_64_centos6/Dockerfile [hunk 7]
```diff
@@ -15,15 +17,12 @@ RUN yum update -y && yum install -y \
...
 
-RUN curl -sSL https://get.rvm.io | /bin/bash -s stable
```
Comment (igorpeshansky) [inline build/x86_64_centos6/Dockerfile]: Can we do this for the centos7 Dockerfile as well?

File: build/x86_64_debian/Dockerfile [hunk 8]
```diff
@@ -11,7 +11,7 @@ RUN /bin/bash -l -c "apt-get update -y && apt-get install -y \
         curl \
   && gpg --keyserver hkp://keys.gnupg.net --recv-keys 409B6B1796C275462A1703113804BB82D39DC0E3 \
   && curl -sSL https://get.rvm.io | /bin/bash -s stable \
```
Comment (igorpeshansky) [inline build/x86_64_debian/Dockerfile]: Let's move this to just before the "rvm" command. Also see my other comment about the scope of the pipe.

File: build/x86_64_centos6/Dockerfile [hunk 9]
```diff
@@ -15,15 +17,12 @@ RUN yum update -y && yum install -y \
...
 RUN /bin/bash -l -c "rvm requirements && rvm install 2.4 && gem install bundler --no-ri --no-rdoc && gem update"
```
Comment (igorpeshansky) [inline build/x86_64_centos6/Dockerfile]: We should be able to fold this `bash -l` command into the previous `RUN` as well.

File: build/x86_64_centos6/Dockerfile (hunk 6 above)
Comment (dhrupadb) [reply build/x86_64_centos6/Dockerfile]: Done.

File: build/x86_64_centos6/Dockerfile (hunk 7 above)
Comment (dhrupadb) [reply build/x86_64_centos6/Dockerfile]: Done

File: build/x86_64_centos6/Dockerfile (hunk 9 above)
Comment (dhrupadb) [reply build/x86_64_centos6/Dockerfile]: Done.

File: build/x86_64_debian/Dockerfile (hunk 8 above)
Comment (dhrupadb) [reply build/x86_64_debian/Dockerfile]: Done.

File: build/x86_64_debian/Dockerfile [hunk 10]
```diff
@@ -10,8 +10,7 @@ RUN apt-get update -y && apt-get install -y \
         fakeroot \
         curl \
   && gpg --keyserver hkp://keys.gnupg.net --recv-keys 409B6B1796C275462A1703113804BB82D39DC0E3 \
-  && curl -sSL https://get.rvm.io | /bin/bash -s stable \
+  && (curl -sSL https://get.rvm.io | /bin/bash -s stable) \
```
Comment (igorpeshansky) [inline build/x86_64_debian/Dockerfile]: Let's move this after the `git config` commands, for consistency.
In general, let's make sure the commands are executed in the same order in all 3 `Dockerfile`s, or that there's a comment explaining why the order is different.

Comment (dhrupadb) [reply build/x86_64_debian/Dockerfile]: Done

File: build/x86_64_centos7/Dockerfile [hunk 11]
```diff
@@ -0,0 +1,25 @@
+FROM centos:7
+
+ENV PATH ${PATH}:/usr/local/bin
+
+RUN yum update -y && yum -y install \
+  autoconf \
```
Comment (bmoyles0117) [inline build/x86_64_centos7/Dockerfile]: This is the last thing that stands out to me. Can we keep the indentation the same as what we have for debian specifically for the packages that are being installed? They should be one level deeper than the rest of the chained commands.

File: build/x86_64_centos6/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,27 @@
+FROM centos:6
+
+ENV PATH ${PATH}:/usr/local/bin
+
+RUN yum update -y && yum install -y \
+  autoconf \
```
Comment (igorpeshansky) [inline build/x86_64_centos6/Dockerfile]: Let's indent the packages, the same way they are indented in the Debian `Dockerfile`. That makes it immediately obvious where the next command starts.

File: build/x86_64_centos6/Dockerfile [hunk 13]
```diff
@@ -0,0 +1,27 @@
+FROM centos:6
+
+ENV PATH ${PATH}:/usr/local/bin
```
Comment (igorpeshansky) [inline build/x86_64_centos6/Dockerfile]: Just curious: why is this not needed on Debian? Is this where `rvm` lives on CentOS?

File: build/x86_64_centos6/Dockerfile (hunk 12 above)
Comment (dhrupadb) [reply build/x86_64_centos6/Dockerfile]: Done.

File: build/x86_64_centos6/Dockerfile (hunk 13 above)
Comment (dhrupadb) [reply build/x86_64_centos6/Dockerfile]: I think it crept in. The build works fine. removed it for now.

File: build/x86_64_centos7/Dockerfile (hunk 11 above)
Comment (dhrupadb) [reply build/x86_64_centos7/Dockerfile]: Done.

**Pull request merged**